torch/_inductor/optimize_indexing.py

```diff
@@ -0,0 +1,508 @@
+        return ValueRanges(-math.inf, math.inf)
+
+    def load(self, name: str, index: sympy.Expr):
+        return ValueRanges(-math.inf, math.inf)
+
+    def store(self, name, index, value, mode=None):
+        return
+
+    def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+        return ValueRanges(-math.inf, math.inf)
+
+    def index_expr(self, index, dtype):
+        assert isinstance(index, ValueRanges)
+        return index
+
+    @staticmethod
+    def to_dtype(x, dtype: torch.dtype):
+        def is_bool(val):
+            return (
+                isinstance(val, bool) or hasattr(low, "is_Boolean") and low.is_Boolean
+            )
+
+        x = ValueRanges.wrap(x)
+        low, up = x.lower, x.upper
+        if is_bool(low):
+            assert is_bool(up)
+            if dtype.is_floating_point:
+                return ValueRanges(sympy.Float(0.0), sympy.Float(1.0))
+            else:
+                return ValueRanges(sympy.Integer(0), sympy.Integer(1))
+        return ValueRanges.wrap(x)
+
+    @staticmethod
+    def constant(value, dtype):
+        # using nan makes subsequent computation throw, and for the purposes of optimization
+        # returning -math.inf - math.inf is equivalent to giving up
+        if math.isnan(value):
+            return ValueRanges(-math.inf, math.inf)
+        if isinstance(value, int):
+            return ValueRanges(sympy.Integer(value), sympy.Integer(value))
+        else:
+            return ValueRanges(sympy.Float(value), sympy.Float(value))
+
+    @staticmethod
+    def reciprocal(x):
+        return ValueRanges.checked_unary_map(x, lambda y: 1 / y)
+
+    @staticmethod
+    def abs(x):
+        return ValueRanges.checked_unary_map(x, abs)
+
+    @staticmethod
+    def neg(x):
+        return ValueRanges.checked_unary_map(x, lambda x: -x)
+
+    @staticmethod
+    def truediv(a, b):
+        return ValueRanges.binary_map_products(a, b, operator.truediv)
+
+    @staticmethod
+    def div(a, b):
+        return ValueRanges.binary_map_products(a, b, operator.truediv)
+
+    @staticmethod
+    def add(a, b):
+        return ValueRanges.binary_map(a, b, operator.add)
+
+    @staticmethod
+    def mul(a, b):
+        return ValueRanges.binary_map_products(a, b, operator.mul)
+
+    @staticmethod
+    def sub(a, b):
+        return ValueRanges.binary_map_products(a, b, operator.sub)
+
+    @staticmethod
+    def exp(x):
+        return ValueRanges.unary_map(x, sympy.functions.elementary.exponential.exp)
+
+    @staticmethod
+    def square(x):
+        return ValueRanges.checked_unary_map(x, lambda y: y * y)
+
+    @staticmethod
+    def log(x):
+        return ValueRanges.checked_unary_map(
+            x, lambda y: -math.inf if y <= 0 else sympy.log(y)
+        )
+
+    @staticmethod
+    def sqrt(x):
+        return ValueRanges.unary_map(x, sympy.sqrt)
+
+    @staticmethod
+    def pow(a, b):
+        return ValueRanges.binary_map_products(a, b, operator.pow)
```

Comment:
I think this one is not correct. Take for example:
```
a = [-3,0]
b = [0,3]
```

Current code will test the cross-product and give: [-27, 0]
However, (-3)**2 = 9

Reply:
Thanks - @lezcano made a number of bug fixes. Check master for up to date pass.